src/devices/Arduino/ArduinoGpioControllerDriver.cs

```diff
@@ -0,0 +1,306 @@
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                throw new NotSupportedException($"Mode {mode} is not supported on Pin {pinNumber}.");
+            }
+
+            _arduinoBoard.Firmata.SetPinMode(pinNumber, firmataMode);
+
+            // Cache this value. Since the GpioController calls GetPinMode when trying to write a pin (to verify whether it is set to output),
+            // that would be very expensive here. And setting output pins should be cheap.
+            _pinModes[pinNumber] = mode;
+        }
+
+        protected override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinModes.TryGetValue(pinNumber, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            SupportedMode mode = _arduinoBoard.Firmata.GetPinMode(pinNumber);
+
+            PinMode ret;
+            switch (mode)
+            {
+                case SupportedMode.DIGITAL_OUTPUT:
+                    ret = PinMode.Output;
+                    break;
+                case SupportedMode.INPUT_PULLUP:
+                    ret = PinMode.InputPullUp;
+                    break;
+                case SupportedMode.DIGITAL_INPUT:
+                    ret = PinMode.Input;
+                    break;
+                default:
+                    ret = PinMode.Input; // TODO: Return "Unknown"
+                    break;
+            }
+
+            _pinModes[pinNumber] = ret;
+            return ret;
+        }
+
+        protected override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            SupportedMode firmataMode;
+            switch (mode)
+            {
+                case PinMode.Output:
+                    firmataMode = SupportedMode.DIGITAL_OUTPUT;
+                    break;
+                case PinMode.InputPullUp:
+                    firmataMode = SupportedMode.INPUT_PULLUP;
+                    break;
+                case PinMode.Input:
+                    firmataMode = SupportedMode.DIGITAL_INPUT;
+                    break;
+                default:
+                    return false;
+            }
+
+            var pinConfig = _supportedPinConfigurations.FirstOrDefault(x => x.Pin == pinNumber);
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected override PinValue Read(int pinNumber)
+        {
+            return _arduinoBoard.Firmata.GetDigitalPinState(pinNumber);
+        }
+
+        protected override void Write(int pinNumber, PinValue value)
+        {
+            _arduinoBoard.Firmata.WriteDigitalPin(pinNumber, value);
+        }
+
+        protected override WaitForEventResult WaitForEvent(int pinNumber, PinEventTypes eventTypes, CancellationToken cancellationToken)
```

Comment:
Any idea of the performance of the events?

Reply:
I didn't take explicit measurements (maybe I should), but probably not to bad, since they're really handled as interrupts and no polling from the client side is required here.

Reply:
> since they're really handled as interrupts and no polling from the client side is required here

Then that's perfect, it means that you can't miss any. Otherwise I would have kind of argued about having them. Now, you can definitely keep them :-)